### DISCUSSION TITLE
New commands to the media player entities

### DISCUSSION THREAD

**Discussion by albaintor** [home-assistant/architecture#1061] [Entity Models] [answered]
Hi all,
I have created the following pull request : https://github.com/home-assistant/core/pull/113147 that has an impact on the entity model indeed.
Here is the context : actually the media player entity is great with features that can be enabled by the integrations based on the capabilities, so that we have an interface that changes wether we are playing music, video or switch TV channels...

But with the multiplication of streaming devices, some common features are missing, which includes :

- Arrow keys navigation
- Raise menu, context menu, home menu
- Number digits to switch channels or go to chapter/media position
- Language selection for audio or subtitles
- Record live TV or access to the recordings

This is possible to implement this feature through the remote entity by sending custom commands but :

1. We can't get benefits of these features in the media player card (not the same entity)
2. There is no standard command names in Home Assistant, so this is not possible to design a generic card that would handle  those commands

My PR brings the following changes :

- Define a list of predefined commands that can be triggered directly from a service call or a card without bothering with the commands of target device : the mapping will be done by the integration between this predefined list and the commands of the device
- Define a list of new features for the media player entity so that the front end know which ones are available or not : for example if the arrow keys are available, the card can display the arrows or else hide them
- Expose the send_command method in the media player entity (which is used at last to send the mapped commands)
- Not done but could be useful also : audio & subtitle language list selection as like source selection

I guess there will be debate why not adding those features to the remote entity instead of the media player : I don't see it as a duplicate as remote entity is still useful and fills its duty (handle commands), however the media player can become a more advanced card but still with simple usage as the UI adapts itself according to the device capabilities.

As real examples that would get benefits of this feature  : Apple TV (menu and dashboard navigation, language selection) , Android TV (same), STB (menus & recordings) and other media streamers, bluray players, TVs... any device with a screen and menu navigation actually.

**Comment by MartinHjelmare** [comment, marked as answer]
We've discussed this proposal in the core team. We've decided to decline the proposal. Here's why:

- We don't add entity services based only on frontend needs, like a media player card. There should be motivation how a user would use the new service in an automation or script.
- We're moving away from complex entities that have many different features and services and represent the whole device in favor of specialized entities that each represent one feature of a device. Multiple entities can be connected to the same Home Assistant device to signal the context.
- We already have a remote entity with a similar service.
- Integrations can add custom entity services for each feature that is needed, that we haven't already standardized.

**Comment by MartinHjelmare** [comment]
Side note: If you're interested in working in this area, I'd suggest looking at how we can improve the remote integration and make it more useful.

**Reply by albaintor** [reply]
Hi,
does it mean that the suggested modifications could be switched to the remote entity ?
To summarize, there are 2 user needs/requirements : 
1/ Having a list of common commands to be used to avoid specific coding. This could be only documentation or enumerations with mapping as in the PR
2/ Getting and selecting in a list provided by the entity : select & switch audio/subtitles, TV channels, radio...

At this state there is no solution to provide a good remote control from HA, I wonder if there is a strategy or thoughts on your side about this
Thanks anyway

**Reply by MartinHjelmare** [reply]
1. I'd look at how the event entity allows setting a list of known event types with translations.
2. That's not easily done if we want translations for the commands, since translations need to be pre-defined.

Please open a new discussion for this so we don't keep talking here. Thanks!

**Reply by albaintor** [reply]
Ok done !
